test/e2e/validation.go

```diff
@@ -22,199 +22,211 @@ import (
+				namespace := framework.CreateTestNamespace(kubeClient, testBaseName)
+				framework.AddCleanupAction(func() {
+					framework.DeleteNamespace(kubeClient, namespace)
+				})
+
+				By(fmt.Sprintf("Creating an invalid %s in the separate test namespace %s", resourceName, namespace))
+				invalidObj := vrt.invalidObject(namespace)
+				err := client.Create(context.TODO(), invalidObj)
+				if err != nil {
+					f.Logger().Fatalf("Unexpected error creating invalid %s = %+v in another test namespace %s, err: %v", resourceName, invalidObj, namespace, err)
+				}
+			})
+		})
+	}
 })
+
+type validationResourceTest interface {
+	initialize() pkgruntime.Object
+	getObjectMeta() *metav1.ObjectMeta
+	invalidObject(namespace string) pkgruntime.Object
+	setInvalidField(obj pkgruntime.Object)
+	validObject() pkgruntime.Object
+	invalidObjectFromValid(obj pkgruntime.Object) pkgruntime.Object
+}
+
+type ftcValidationTest struct {
+	object pkgruntime.Object
+}
+
+type kfclusterValidationTest struct {
+	object pkgruntime.Object
+}
+
+func newValidationResourceTest(resourceName string) validationResourceTest {
+	var vrt validationResourceTest
+	switch resourceName {
+	case ftc.ResourceName:
+		vrt = &ftcValidationTest{}
+	case kfcluster.ResourceName:
+		vrt = &kfclusterValidationTest{}
+	}
+	return vrt
+}
+
+func (ftc *ftcValidationTest) initialize() pkgruntime.Object {
+	if ftc.object != nil {
+		return ftc.object
+	}
+
+	// For the target API type, use an existing K8s API resource that
+	// is not currently enabled by default. This simplifies logic and
+	// avoids having to create a CRD that prevents validation tests
+	// from running with LimitedScope.
+	apiResource := metav1.APIResource{
+		Group:      "apps",
+		Version:    "v1",
+		Kind:       "DaemonSet",
+		Name:       "daemonsets",
+		Namespaced: true,
+	}
+	enableTypeDirective := enable.NewEnableTypeDirective()
+	ftc.object = enable.GenerateTypeConfigForTarget(apiResource, enableTypeDirective).(*v1beta1.FederatedTypeConfig)
+	return ftc.object
+}
+
+func (ftc *ftcValidationTest) getObjectMeta() *metav1.ObjectMeta {
+	return &ftc.object.(*v1beta1.FederatedTypeConfig).ObjectMeta
+}
+
+func (ftc *ftcValidationTest) invalidObject(namespace string) pkgruntime.Object {
+	invalidFtc := ftc.object.DeepCopyObject().(*v1beta1.FederatedTypeConfig)
+	if namespace != "" {
+		invalidFtc.Namespace = namespace
+	}
+
+	ftc.setInvalidField(invalidFtc)
+	return invalidFtc
+}
+
+func (ftc *ftcValidationTest) setInvalidField(obj pkgruntime.Object) {
+	obj.(*v1beta1.FederatedTypeConfig).Spec.FederatedType.Group = ""
+}
+
+func (ftc *ftcValidationTest) validObject() pkgruntime.Object {
+	return ftc.object.DeepCopyObject()
+}
+
+func (ftc *ftcValidationTest) invalidObjectFromValid(obj pkgruntime.Object) pkgruntime.Object {
+	invalidFtc := obj.DeepCopyObject().(*v1beta1.FederatedTypeConfig)
+	ftc.setInvalidField(invalidFtc)
+	return invalidFtc
+}
+
+func (kfc *kfclusterValidationTest) initialize() pkgruntime.Object {
```

Comment:
nit: kfCluster?

Reply:
That's what I originally had and then changed it since it's referred to `kfcluster` in other places such as the import alias. :smile: